File: NetFramework/RulesManager/VB/DataGridViewConditionalFormatting/FormattableDataGridView.vb
Imports System.ComponentModel
Imports C1.Win.CoditionalFormatting

Public Class FormattableDataGridView
    Inherits DataGridView
    Implements IFormattableView

    Public Event DataChanged As ListChangedEventHandler Implements IFormattableView.DataChanged
    Public Event ItemFormatting As EventHandler(Of ItemFormattingEventArgs) Implements IFormattableView.ItemFormatting

    Private Function GetFieldNames() As IEnumerable(Of String) Implements IFormattableView.GetFieldNames
        Dim fields = New List(Of String)()

        For Each column In Columns
            Dim name = (TryCast(column, DataGridViewColumn))?.DataPropertyName

            If Not String.IsNullOrEmpty(name) Then
                fields.Add(name)
            End If
        Next

        Return fields
    End Function

    Private Function GetItemCount() As Integer Implements IFormattableView.GetItemCount
        Return Rows.Count
    End Function

    Private Function GetValue(row As Integer, col As String) As Object Implements IFormattableView.GetValue
        Return Me(col, row).Value
    End Function

    Protected Overrides Sub OnDataBindingComplete(ByVal e As DataGridViewBindingCompleteEventArgs)
        MyBase.OnDataBindingComplete(e)
        RaiseEvent DataChanged(Me, New ListChangedEventArgs(ListChangedType.Reset, 0))
    End Sub

    Protected Overrides Sub OnCellPainting(ByVal e As DataGridViewCellPaintingEventArgs)
        Dim columnIndex = e.ColumnIndex
        Dim rowIndex = e.RowIndex

        If columnIndex >= 0 AndAlso rowIndex >= 0 Then
            Dim ifea = New ItemFormattingEventArgs(rowIndex, Columns(columnIndex).DataPropertyName)
            RaiseEvent ItemFormatting(Me, ifea)
            Dim ruleStyle = ifea?.Style

            If ruleStyle IsNot Nothing Then
                Dim cellStyle = e.CellStyle
                Dim backColor = ruleStyle.BackColor

                If Not backColor.IsEmpty Then
                    cellStyle.BackColor = backColor
                End If

                Dim foreColor = ruleStyle.ForeColor

                If Not foreColor.IsEmpty Then
                    cellStyle.ForeColor = foreColor
                End If

                Dim fontStyle = ruleStyle.FontStyle

                If cellStyle.Font.Style <> fontStyle AndAlso fontStyle <> FontStyle.Regular Then
                    cellStyle.Font = New Font(cellStyle.Font, fontStyle)
                End If
            End If
        End If

        MyBase.OnCellPainting(e)
    End Sub

    Public Sub UpdateView() Implements IFormattableView.UpdateView
        Invalidate()
    End Sub

End Class

File: NetFramework/RulesManager/CS/DataGridViewConditionalFormatting/FormattableDataGridView.cs
using C1.Win.CoditionalFormatting;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using System.ComponentModel;

namespace DataGridViewConditionalFormatting
{
    public class FormattableDataGridView : DataGridView, IFormattableView
    {
        public event ListChangedEventHandler DataChanged;
        public event EventHandler<ItemFormattingEventArgs> ItemFormatting;

        IEnumerable<string> IFormattableView.GetFieldNames()
        {
            var fields = new List<string>();
            foreach (var column in Columns)
            {
                var name = (column as DataGridViewColumn)?.DataPropertyName;
                if (!string.IsNullOrEmpty(name))
                {
                    fields.Add(name);
                }
            }

            return fields;
        }

        int IFormattableView.GetItemCount()
        {
            return Rows.Count;
        }

        object IFormattableView.GetValue(int row, string col)
        {
            return this[col, row].Value;
        }

        protected override void OnDataBindingComplete(DataGridViewBindingCompleteEventArgs e)
        {
            base.OnDataBindingComplete(e);
            DataChanged?.Invoke(this, new ListChangedEventArgs(ListChangedType.Reset, 0));
        }

        protected override void OnCellPainting(DataGridViewCellPaintingEventArgs e)
        {
            var columnIndex = e.ColumnIndex;
            var rowIndex = e.RowIndex;
            if (columnIndex >= 0 && rowIndex >= 0)
            {
                var ifea = new ItemFormattingEventArgs(rowIndex, Columns[columnIndex].DataPropertyName);
                ItemFormatting?.Invoke(this, ifea);

                var ruleStyle = ifea?.Style;
                if (ruleStyle != null)
                {
                    var cellStyle = e.CellStyle;

                    var backColor = ruleStyle.BackColor;
                    if (!backColor.IsEmpty)
                    {
                        cellStyle.BackColor = backColor;
                    }

                    var foreColor = ruleStyle.ForeColor;
                    if (!foreColor.IsEmpty)
                    {
                        cellStyle.ForeColor = foreColor;
                    }

                    var fontStyle = ruleStyle.FontStyle;
                    if (cellStyle.Font.Style != fontStyle && fontStyle != FontStyle.Regular)
                    {
                        cellStyle.Font = new Font(cellStyle.Font, fontStyle);
                    }
                }
            }

            base.OnCellPainting(e);
        }

        void IFormattableView.UpdateView()
        {
            Invalidate();
        }
    }
}
